src/lib.rs

```diff
@@ -205,4 +227,39 @@ mod tests {
         println!("{}", &str);
         assert_eq!(String::from("string"), str.to_string())
     }
+
+    #[test]
+    fn from_str_works() {
+        let valid_str = "string";
+
+        let _non_empty_string = NonEmptyString::from_str("").expect_err("operation must be failed");
+
+        let non_empty_string = NonEmptyString::from_str(valid_str).unwrap();
+        assert_eq!(non_empty_string.as_str(), valid_str);
+        assert_eq!(non_empty_string, valid_str.parse().unwrap());
+    }
+
+    #[test]
+    fn into_works() {
+        let non_empty_string = NonEmptyString::new("string".to_string()).unwrap();
+        let _string: String = non_empty_string.into();
+
+        let non_empty_string = NonEmptyString::new("string".to_string()).unwrap();
+        let _string = String::from(non_empty_string);
+    }
+
+    #[test]
+    fn hash_works() {
+        let mut map = HashMap::new();
+        map.insert(NonEmptyString::from_str("id.1").unwrap(), 1);
+        map.insert(NonEmptyString::from_str("id.2").unwrap(), 2);
+
+        assert_eq!(map.len(), 2);
+
+        let mut set = HashSet::new();
+        set.insert(NonEmptyString::from_str("1").unwrap());
+        set.insert(NonEmptyString::from_str("2").unwrap());
+
+        assert_eq!(set.len(), 2);
+    }
```

Comment:
I don't feel there is any need to test the derived `Hash` attribute. If we want to solidify the fact that it implements `Hash`, the test could be constructed with a helper function like below:
```rust
fn is_hash<T: Hash>() {}

#[test]
fn non_empty_string_impl_hash() {
    is_hash::<NonEmptyString>()
}
```